### ISSUE TITLE
KVM driver fails when --forward-ports set to true

### ISSUE THREAD

**Issue by marekjelen** [minishift/minishift#512]
Hi,

KVM seems to be failing on F25, my configuration is

```
export MINISHIFT_VM_DRIVER="kvm"
export MINISHIFT_MEMORY="4096"
export MINISHIFT_OPENSHIFT_VERSION="v1.5.0-alpha.3"
export MINISHIFT_FORWARD_PORTS="true"
```

and running `minishift start` fails on

```
Starting local OpenShift cluster using 'kvm' hypervisor...
 19.93 MB / 19.93 MB [=================================================================================================================================================================================] 100.00% 0s
Provisioning OpenShift via '/home/mjelen/.minishift/cache/oc/v1.5.0-alpha.3/oc [cluster up --use-existing-config --forward-ports true --host-config-dir /var/lib/minishift/openshift.local.config --host-data-dir /var/lib/minishift/hostdata --host-volumes-dir /var/lib/minishift/openshift.local.volumes]'
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.0-alpha.3 image ... 
   Pulling image openshift/origin:v1.5.0-alpha.3
   Pulled 0/3 layers, 3% complete
   Pulled 1/3 layers, 35% complete
   Pulled 2/3 layers, 90% complete
   Pulled 3/3 layers, 100% complete
   Extracting
   Image pull complete
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 127.0.0.1 as the server IP
-- Starting OpenShift container ... 
   Creating initial OpenShift configuration
   Starting OpenShift using container 'origin'
   Waiting for API server to start listening
FAIL
   Error: timed out waiting for OpenShift container "origin" 
   WARNING: 127.0.0.1:8443 may be blocked by firewall rules
   Details:
     Last 10 lines of "origin" container log:
     E0303 10:36:17.717355    3134 util.go:45] Metric for replenishment_controller already registered
     I0303 10:36:17.792665    3134 secret_updating_controller.go:108] starting service signing cert update controller
     I0303 10:36:17.814498    3134 start_master.go:762] Started Origin Controllers
     E0303 10:36:17.842547    3134 actual_state_of_world.go:462] Failed to set statusUpdateNeeded to needed true because nodeName="192.168.122.215"  does not exist
     I0303 10:36:17.862501    3134 nodecontroller.go:429] Initializing eviction metric for zone: 
     W0303 10:36:17.862581    3134 nodecontroller.go:678] Missing timestamp for Node 192.168.122.215. Assuming now as a timestamp.
     I0303 10:36:17.862673    3134 nodecontroller.go:608] NodeController detected that zone  is now in state Normal.
     I0303 10:36:17.862729    3134 event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"192.168.122.215", UID:"3ab39c4c-fffd-11e6-a2df-6e051db270a2", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node 192.168.122.215 event: Registered Node 192.168.122.215 in NodeController
     W0303 10:36:22.707305    3134 lease_endpoint_reconciler.go:174] Resetting endpoints for master service "kubernetes" to [192.168.122.215]
     2017-03-03 10:36:37.116060 W | wal: sync duration of 1.121638504s, expected less than 1s


   Solution:
     Ensure that you can access 127.0.0.1:8443 from your machine
E0303 11:36:52.059843   14548 start.go:370] Error starting the cluster:  exit status 1

```

**Comment by praveenkumar** [comment]
That looks like firewall issue can you open "8443" port for localhost (using firewalld) and try again? May be recently fedora made changes against firewalld.

**Comment by LalatenduMohanty** [comment]
The default `minishift start` works for me on F25. I will try the exact settings mentioned in the report.

**Comment by hferentschik** [comment]
@marekjelen and that's F25 specific? Aka it works on other Fedora versions? What's about another driver? Basically I am wondering whether this is an issue related to KVM / Fedora or the fact that you use port forwards.

**Comment by marekjelen** [comment]
@hferentschik do not know, I have only F25 set up.

@praveenkumar tried enabling net.ipv4.ip_forward and adding port 8443/tcp to firewalld public zone

**Comment by praveenkumar** [comment]
@marekjelen can you unset all your shell variable and the just try `openshift start` by default it use kvm driver.

**Comment by marekjelen** [comment]
@praveenkumar 

```
Starting local OpenShift cluster using 'kvm' hypervisor...
 22.37 MB / 22.37 MB [=================================================================================================================================================================================] 100.00% 0s
Provisioning OpenShift via '/home/mjelen/.minishift/cache/oc/v1.4.1/oc [cluster up --use-existing-config --host-config-dir /var/lib/minishift/openshift.local.config --host-data-dir /var/lib/minishift/hostdata --host-volumes-dir /var/lib/minishift/openshift.local.volumes]'
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... 
   Deleted existing OpenShift container
-- Checking for openshift/origin:v1.4.1 image ... 
   Pulling image openshift/origin:v1.4.1
   Pulled 1/3 layers, 37% complete
   Pulled 2/3 layers, 69% complete
   Pulled 3/3 layers, 100% complete
   Extracting
   Image pull complete
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 192.168.42.154 as the server IP
-- Starting OpenShift container ... FAIL
   Error: Docker run error rc=2
   Details:
     Image: openshift/origin:v1.4.1
     Entrypoint: [/bin/bash]
     Command: [-c for name in 192.168.42.154 minishift; do ls /var/lib/origin/openshift.local.config/node-$name &> /dev/null && echo $name && break; done]
E0303 14:17:48.381628   19534 start.go:370] Error starting the cluster:  exit status 1
```

**Comment by praveenkumar** [comment]
@marekjelen Now we got something different, can you follow below steps and let me know if still fails?

```
# ./minishift delete
# rm -fr ~/.minishift
# ./minishift start --memory 4096 --openshift-version v1.5.0-alpha.3
```

**Comment by marekjelen** [comment]
@praveenkumar yes, that worked

```
Starting local OpenShift cluster using 'kvm' hypervisor...
Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.1/minishift-b2d.iso'
 39.00 MB / 39.00 MB [=================================================================================================================================================================================] 100.00% 0s
 19.93 MB / 19.93 MB [=================================================================================================================================================================================] 100.00% 0s
Provisioning OpenShift via '/home/mjelen/.minishift/cache/oc/v1.5.0-alpha.3/oc [cluster up --use-existing-config --host-config-dir /var/lib/minishift/openshift.local.config --host-data-dir /var/lib/minishift/hostdata --host-volumes-dir /var/lib/minishift/openshift.local.volumes]'
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.0-alpha.3 image ... 
   Pulling image openshift/origin:v1.5.0-alpha.3
   Pulled 0/3 layers, 3% complete
   Pulled 1/3 layers, 37% complete
   Pulled 2/3 layers, 87% complete
   Pulled 3/3 layers, 100% complete
   Extracting
   Image pull complete
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 192.168.42.141 as the server IP
-- Starting OpenShift container ... 
   Creating initial OpenShift configuration
   Starting OpenShift using container 'origin'
   Waiting for API server to start listening
   OpenShift server started
-- Adding default OAuthClient redirect URIs ... OK
-- Installing registry ... OK
-- Installing router ... OK
-- Importing image streams ... OK
-- Importing templates ... OK
-- Login to server ... OK
-- Creating initial project "myproject" ... OK
-- Removing temporary directory ... OK
-- Checking container networking ... OK
-- Server Information ... 
   OpenShift server started.
   The server is accessible via web console at:
       https://192.168.42.141:8443

   You are logged in as:
       User:     developer
       Password: developer

   To login as administrator:
       oc login -u system:admin
```

**Comment by praveenkumar** [comment]
@marekjelen your issue was with port forward nothing else and second time you got different result because first time run made minishift home profile dirty so once it cleaned you got expected result.

**Comment by marekjelen** [comment]
@praveenkumar awesome, wiped all the stuff as you suggested; set the env vars except port forward; rerun and works fine

@hferentschik looks like the problem is the port forwarding :)

**Comment by praveenkumar** [comment]
@marekjelen Thanks you for trying something different :)

**Comment by LalatenduMohanty** [comment]
I can reproduce the issue on Fedora 25 

```
$ env | grep MINI
MINISHIFT_FORWARD_PORTS=true


$ minishift start
Starting local OpenShift cluster using 'kvm' hypervisor...
Provisioning OpenShift via '/home/lmohanty/.minishift/cache/oc/v1.4.1/oc [cluster up --use-existing-config --forward-ports true --host-config-dir /var/lib/minishift/openshift.local.config --host-data-dir /var/lib/minishift/hostdata --host-volumes-dir /var/lib/minishift/openshift.local.volumes]'
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.4.1 image ... 
   Pulling image openshift/origin:v1.4.1
   Pulled 0/3 layers, 0% complete
   Pulled 1/3 layers, 33% complete
   Pulled 2/3 layers, 67% complete
   Pulled 3/3 layers, 100% complete
   Extracting
   Image pull complete
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 127.0.0.1 as the server IP
-- Starting OpenShift container ... FAIL
   Error: exec: "socat": executable file not found in $PATH
E0303 19:21:30.679470    1504 start.go:386] Error starting the cluster:  exit status 1
```

**Comment by jorgemoralespou** [comment]
@LalatenduMohanty @hferentschik for linux os (which is always the case) port forward should always be false. I think that this option should be silently swallowed by minishift in case the user specifies it. 

cc/ @csrwng

**Comment by jorgemoralespou** [comment]
Here. 
The second error that @marekjelen had was because the VM got a different ip. This is s known issue in oc cluster. I have described this in another issue. 
If the vm can not always get the same ip once created, we're going to run into issues until 1.6 of oc. This might be specific to the KVM driver here. 

cc/ @csrwng another occurrence of this issue

**Comment by praveenkumar** [comment]
@jorgemoralespou If that is the case shouldn't we put it in the docs as known issue for kvm, port forward not allowed, @hferentschik  wdyt?

**Comment by LalatenduMohanty** [comment]
@jorgemoralespou IMO  in that case the best thing would be let user know when the `minishift start` gets executed  and till we have that , we must document it.

**Comment by jorgemoralespou** [comment]
@LalatenduMohanty IMHO, since "oc cluster" is made to support multiple OS, but in minishift it will always be run on the same OS, which happens to be linux, and that flag is not meant for linux, I would try to avoid user errors, rather than given them feedback to fix their errors. But that is the approach I would take. 

For the KVM issue, yes, I agree that is a know issue, but it's not really an issue but a bug in the driver, so I'd rather fix it than document it.

**Comment by praveenkumar** [comment]
> IMHO, since "oc cluster" is made to support multiple OS, but in minishift it will always be run on the same OS, which happens to be linux, and that flag is not meant for linux, I would try to avoid user errors, rather than given them feedback to fix their errors. But that is the approach I would take.

That make sense now so actually we should hide that flag for the minishift so user don't even set it.

> For the KVM issue, yes, I agree that is a know issue, but it's not really an issue but a bug in the driver, so I'd rather fix it than document it.

What is the KVM issue for the driver?

**Comment by jorgemoralespou** [comment]
@praveenkumar KVM driver should provide the same IP to the VM for every start of the KVM VM. You should be able to verify by starting and stopping several times minishift using KVM and verifying which IP you get for the VM.
I think the minishift drivers should properly configure the networking like xhyve, virtualbox does. Every restart of an already created VM should provide always the same IP, no matter what network you're attached to.

**Comment by praveenkumar** [comment]
@jorgemoralespou Ack, will test it out and look into it.

**Comment by hferentschik** [comment]
@praveenkumar is there an update on this?

**Comment by praveenkumar** [comment]
> Is there an update on this?

yes I did try with kvm and as @jorgemoralespou said it always give a different IP from the network pool which is not something in our hand even I have updated kvm-driver but this issue is actually hide `--forward-port` from minishift side which I will work on.

**Comment by hferentschik** [comment]
> this issue is actually hide --forward-port from minishift side which I will work on

So you are suggesting to remove this option altogether? Does it not work with other drivers which don't change the IP? Or are we saying forward-port does never work in the context of Minishift?

**Comment by praveenkumar** [comment]
> Or are we saying forward-port does never work in the context of Minishift?

Yes forward-port does never work in the context of Minishift.

**Comment by hferentschik** [comment]
Because port forward it towards the VM and not to the host, right?

**Comment by praveenkumar** [comment]
> Because port forward it towards the VM and not to the host, right?

Right and for us it always the Linux VM were this not works as @jorgemoralespou described.

**Comment by hferentschik** [comment]
+1 for not exposing the option

**Comment by praveenkumar** [comment]
> Because port forward it towards the VM and not to the host, right?

That's right.

**Comment by praveenkumar** [comment]
> Because port forward it towards the VM and not to the host, right?

That's right.

**Comment by praveenkumar** [comment]
Fixed by #915
